### PR TITLE
Add time sync functionality, to move the view of time of a remote cluster

### PR DESCRIPTION
**Pull request by wxing1292**
* Add time sync functionality, to move the view of time of a remote cluster

solve #951 #786

### CODE REVIEW

File: service/history/handler.go [hunk 1]
```diff
@@ -924,6 +924,30 @@ func (h *Handler) ReplicateEvents(ctx context.Context, replicateRequest *hist.Re
...
+// SyncShardStatus is called by processor to sync history shrad information from another cluster
+func (h *Handler) SyncShardStatus(ctx context.Context, syncShardStatusRequest *hist.SyncShardStatusRequest) error {
+	h.startWG.Wait()
+
+	h.metricsClient.IncCounter(metrics.HistorySyncShardStatusScope, metrics.CadenceRequests)
+	sw := h.metricsClient.StartTimer(metrics.HistorySyncShardStatusScope, metrics.CadenceLatency)
+	defer sw.Stop()
+
+	// shard ID is already provided in the request
+	engine, err := h.controller.getEngineForShard(int(syncShardStatusRequest.GetShardId()))
```
Comment (samarabbas) [inline service/history/handler.go]: Can you validate if the request has shardID and return BadRequest if it does not.

File: service/history/historyEngine.go [hunk 2]
```diff
@@ -2194,6 +2194,16 @@ func (e *historyEngineImpl) ReplicateEvents(replicateRequest *h.ReplicateEventsR
 	return e.replicator.ApplyEvents(replicateRequest)
 }
 
+func (e *historyEngineImpl) SyncShardStatus(ctx context.Context, request *h.SyncShardStatusRequest) error {
+	clusterName := request.GetSourceCluster()
+	now := time.Unix(0, request.GetTimestamp())
+
+	e.shard.SetCurrentTime(clusterName, now)
+	e.txProcessor.NotifyNewTask(clusterName, now, []persistence.Task{})
```
Comment (samarabbas) [inline service/history/historyEngine.go]: Do we really need this notification here?  I think we already have protection to trigger scan at regular bases.  Firing a notification to force a scan sounds redundant.

File: service/history/shardContext.go [hunk 3]
```diff
@@ -585,6 +589,19 @@ func (s *shardContextImpl) updateShardInfoLocked() error {
 	}
 	updatedShardInfo := copyShardInfo(s.shardInfo)
 
+	if s.messageProducer != nil {
```
Comment (samarabbas) [inline service/history/shardContext.go]: Can we only create sync task on idle scenario?

File: service/history/handler.go (hunk 1 above)
Comment (wxing1292) [reply service/history/handler.go]: sure

File: service/history/queueAckMgr.go [hunk 4]
```diff
@@ -188,7 +188,6 @@ MoveAckLevelLoop:
 			break MoveAckLevelLoop
 		}
 	}
-	updateShard := a.ackLevel != ackLevel
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/queueAckMgr.go]: we have a check on the actual update shard API call, which has min update interval = 5min

File: service/history/shardContext.go (hunk 3 above)
Comment (wxing1292) [reply service/history/shardContext.go]: #956

File: service/history/historyEngine.go (hunk 2 above)
Comment (wxing1292) [reply service/history/historyEngine.go]: talked offline, this should be fine

**Pull request merged**